lib/charms/opensearch/v0/opensearch_base_charm.py

```diff
@@ -603,40 +607,49 @@ def _on_update_status(self, event: UpdateStatusEvent):
 
     def _on_config_changed(self, event: ConfigChangedEvent):  # noqa C901
         """On config changed event. Useful for IP changes or for user provided config changes."""
+        restart_requested = False
         if self.opensearch_config.update_host_if_needed():
+            restart_requested = True
+
             self.status.set(MaintenanceStatus(TLSNewCertsRequested))
             self._delete_stored_tls_resources()
             self.tls.request_new_unit_certificates()
 
             # since when an IP change happens, "_on_peer_relation_joined" won't be called,
             # we need to alert the leader that it must recompute the node roles for any unit whose
             # roles were changed while the current unit was cut-off from the rest of the network
-            self._on_peer_relation_joined(event)
+            self._on_peer_relation_joined(
+                RelationJoinedEvent(event.handle, PeerRelationName, self.app, self.unit)
+            )
 
         previous_deployment_desc = self.opensearch_peer_cm.deployment_desc()
         if self.unit.is_leader():
             # run peer cluster manager processing
+            # todo add check here if the diff can be known from now on already
             self.opensearch_peer_cm.run()
 
             # handle cluster change to main-orchestrator (i.e: init_hold: true -> false)
             self._handle_change_to_main_orchestrator_if_needed(event, previous_deployment_desc)
-        elif not previous_deployment_desc:
-            # deployment desc not initialized yet by leader
-            event.defer()
+
+        # todo: handle gracefully configuration setting at start of the charm
+        if not self.plugin_manager.check_plugin_manager_ready():
             return
 
         try:
             if not self.plugin_manager.check_plugin_manager_ready():
                 raise OpenSearchNotFullyReadyError()
+
             if self.unit.is_leader():
                 self.status.set(MaintenanceStatus(PluginConfigCheck), app=True)
-            if self.plugin_manager.run():
+
+            if self.plugin_manager.run() and not restart_requested:
                 if self.upgrade_in_progress:
                     logger.warning(
                         "Changing config during an upgrade is not supported. The charm may be in a broken, unrecoverable state"
                     )
                     event.defer()
                     return
+
```

Comment:
We should move any "plugin_manager / defer" *before* the line 629:
```
self.opensearch_peer_cm.run()
```
As that `run()` will make a call to `trigger_peer_rel_changed`

Reply:
Please, also a very big comment banner informing: "no defer after this line", with an explanation on *why* we need that precaution.

Reply:
Not sure I understand what you mean exactly here.

> As that run() will make a call to trigger_peer_rel_changed

No - this will only run if there was a change to the 2 config options of the app:
- init-hold
- roles 
You can see this on the logic of `self.opensearch_peer_cm.run()` - it only concerns `exisiting clusters` which we're updating